### PR TITLE
More link shape properties

### LINKED ISSUE #60: Add Additional Link Shape Properties
## Elevator Pitch

Add `line_dash` and `curvature` to the `links`.

## Motivation

* `curvature` and `line dashes` could be used to convey information about relationships.
* would produce a more complete 1:1 with the `force-graph` API.

## Design Ideas
* For curvature it would be pretty straight forward from the `python` side.
  ```python
  curvature: TNumFeature = _make_trait("the curvature of the link, 0: straight, 1: circular", numeric=True)
  ```
* The `line_dash` presents a more complicated challenge, as the data is really a list of numbers, as per the [`HTML Canvas setLineDash`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/setLineDash) method.
  ```python
  line_dash: TListNumFeature = _make_trait(
      "the dash pattern, e.g., [5, 15] to draw a repeating pattern of a 5 pixel segment followed by a 15 pixel blank",
      numeric=True,
      is_list=True,
  )
  ```

### PR DESCRIPTION
**Pull request by sanbales**
## References

Resolves #60 by adding `curvature` and `line dashes` to `LinkShape`.

## Code changes

- [x] add `curvature` to `LinkShape`
- [x] add `line_dash` to `LinkShape`
  - [x] disable this behavior in 3D (it is not supported)
- [x] add coercion to `array` for `line_dash` data
- [x] update notebooks
  - [x] fixed a small issue with `node` selection in `Behaviors.ipynb`
- [x] update docs
- [x] add tests
  - [x] `str` inputs
  - [x] `int` inputs
  - [x] `float` inputs
  - [x] `Column` inputs
  - [x] `Nunjucks` inputs
  - [x] `LinkShape.color`
  - [x] `LinkShape.curvature`
  - [x] `LinkShape.line_dash`
  - [x] `LinkShape.width`
- [x] Interaction with selection
  - [x] need to make sure `LinkSelection` is included before `LinkShape` to show selection styling
  - [x] `LinkSelection` works but it is not tested in conjunction with `LinkShape`
  - [x] future work would be to figure out a way to make `default_curvature=None` to leave the link with its non-selected `curvature`

## User-facing changes

* Two new properties on `LinkShape`:
  * `LinkShape.curvature`:
    * 0.0 produces a straight link (default)
    * 1.0 circular link (i.e., `radius of curvature` = `0.5 * link length`)
    * negative numbers mirror the link
  * `LinkShape.line_dashes`:
    * Define the line dashes using an array of numbers, [reference](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/setLineDash)

image

## Backwards-incompatible changes

> No backwards incompatible changes (**knock on wood**)

### CODE REVIEW

**Comment by sanbales** [comment]
Curvature is no sweat. 2197523bf5659958a6a1ab767ccbbc0147eaaa70 adds `curvature`.

As a `number`:
image

As a `nunjucks` template:
image

**Comment by sanbales** [comment]
`line dash` seems to work as well

[c82314a](https://github.com/jupyrdf/ipyforcegraph/pull/62/commits/c82314a2ef165372212b6fc5a089533132cf9b14)

**Comment by sanbales** [comment]
`3D` has `rotation` for links (spin around the axis between `source` and `target`), not sure it is worth it to bring that in or not, but it could make sense as `dash_line` only works on `2D`.

File: js/utils.ts [hunk 1]
```diff
@@ -32,12 +32,29 @@ export function coerceNumber(value: string): number {
   return Number(value);
 }
 
+export function coerceArray(value: string): any[] {
+  var new_value: any;
+  try {
+    new_value = JSON.parse(value);
+  } catch (e) {
+    console.debug('could not parse ', value, ' as JSON');
```
[the remark this answers is not in the document]
Comment (nrbgt) [reply js/utils.ts:40]: no need to block on this, though

**Review by nrbgt — APPROVED** [review]

**Comment by nrbgt** [comment]
> but it could make sense as dash_line only works on 2D.

Again, not blocking, but should probably filter that out of the 3d demo.

**Comment by sanbales** [comment]
> > but it could make sense as dash_line only works on 2D.
> 
> Again, not blocking, but should probably filter that out of the 3d demo.

Good call, fixed in [da96020](https://github.com/jupyrdf/ipyforcegraph/pull/62/commits/da960202c6085f835e48c70ed698058d4d18e87c)

File: js/utils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (sanbales) [reply js/utils.ts]: Good call, fixed in [dbac506](https://github.com/jupyrdf/ipyforcegraph/pull/62/commits/dbac50669898286bad61f22cefce897ef9bbddff)

**Pull request merged**